File: Common/src/main/java/com/hypherionmc/sdlink/util/translations/SDText.java
package com.hypherionmc.sdlink.util.translations;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author HypherionSA
 * Extended Helper Class to help resolve translation keys into Strings
 */
public final class SDText implements CharSequence {

    private final String translatedText;

    private SDText(String key, @Nullable Object... args) {
        this.translatedText = (args == null || args.length == 0) ?
                TranslationManager.INSTANCE.translate(key)
                : String.format(TranslationManager.INSTANCE.translate(key), args);
    }

    /**
     * Load a translation key into a string
     *
     * @param key The Translation key to load
     * @return The final text, or the translation key if the text does not exist
     */
    public static SDText translate(@NotNull String key) {
        return new SDText(key);
    }

    /**
     * Load a translation key into a string
     *
     * @param key The Translation key to load
     * @param args List of objects that will replace placeholders in translated text
     * @return The final text, or the translation key if the text does not exist
     */
    public static SDText translate(@NotNull String key, @NotNull Object... args) {
        return new SDText(key, args);
    }

    @Override
    public int length() {
        return translatedText.length();
    }

    @Override
    public char charAt(int index) {
        return translatedText.charAt(index);
    }

    @NotNull
    @Override
    public CharSequence subSequence(int start, int end) {
        return translatedText.subSequence(start, end);
    }

    @NotNull
    @Override
    public String toString() {
        return translatedText;
    }
}

File: Common/src/main/kotlin/com/hypherionmc/sdlinkrw/modules/translations/SDText.kt
package com.hypherionmc.sdlinkrw.modules.translations


/**
 * @author HypherionSA
 *
 * Extended Helper Class to help resolve translation keys into Strings
 */
class SDText private constructor(key: String, vararg args: Any?) : CharSequence {

    private val translatedText: String = if (args == null || args.isEmpty())
        TranslationManager.translate(key) else String.format(TranslationManager.translate(key), *args)

    override fun get(index: Int): Char {
        return translatedText[index]
    }

    override val length: Int get() = translatedText.length

    override fun subSequence(startIndex: Int, endIndex: Int): CharSequence {
        return translatedText.subSequence(startIndex, endIndex)
    }

    override fun toString(): String {
        return translatedText
    }

    companion object {
        /**
         * Load a translation key into a string
         *
         * @param key The Translation key to load
         * @return The final text, or the translation key if the text does not exist
         */
        @JvmStatic
        fun translate(key: String): SDText {
            return SDText(key)
        }

        /**
         * Load a translation key into a string
         *
         * @param key The Translation key to load
         * @param args List of objects that will replace placeholders in translated text
         * @return The final text, or the translation key if the text does not exist
         */
        @JvmStatic
        fun translate(key: String, vararg args: Any): SDText {
            return SDText(key, *args)
        }
    }
}